src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -111,11 +117,23 @@ public partial class JsonTypeInfo<T>
                         Utf8JsonWriterCache.ReturnWriter(writer);
                     }
 
-                    await serializationContext.FlushAsync(cancellationToken).ConfigureAwait(false);
+                    FlushResult result = await pipeWriter.FlushAsync(cancellationToken).ConfigureAwait(false);
+                    if (result.IsCanceled || result.IsCompleted)
+                    {
+                        if (result.IsCanceled)
+                        {
+                            ThrowHelper.ThrowOperationCanceledException_PipeWriteCanceled();
+                        }
+
+                        ThrowHelper.ThrowOperationCanceledException_PipeWriteCompleted();
+                    }
                 }
                 finally
                 {
-                    serializationContext.Dispose();
+                    if (pipeWriter is IDisposable disposable)
```

Comment:
I just occurred to me that this would be undesirable in the hypothetical case where you have a user-defined `PipeWriter` that happens to implement `IDisposable`. This should instead check for `PooledByteBufferWriter` (or have the method accept a secondary `IDisposable?` parameter that is only populated by the stream overload.